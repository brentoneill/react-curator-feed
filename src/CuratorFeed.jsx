import React, { Component, Fragment } from "react";
import axios from "axios";
import { format } from "date-fns";

const apiUrl = "https://api.curator.io/v1";

class App extends Component {
  feedId: null;
  posts: null;

  constructor(props) {
    super(props);

    this.state = {
      posts: []
    };
  }

  componentDidMount() {
    const { config } = this.props;

    config.apiUrl = apiUrl ? apiUrl : config.apiUrl;

    axios
      .get(`${config.apiUrl}/feeds?api_key=${config.apiKey}`)
      .then(res => {
        if (res && res.status === 200) {
          this.feedId = res.data[0].id;
          axios
            .get(
              `${config.apiUrl}/feeds/${this.feedId}/posts?api_key=${
                config.apiKey
              }`
            )
            .then(res => {
              if (res && res.status === 200) {
                const { posts } = res.data;
                this.setState({ posts: posts.splice(0, config.postsToShow) });
              }
            })
            .catch(err => {
              console.error(err);
            });
        } else {
          console.log("there was an error getting feed");
        }
      })
      .catch(err => {
        console.error(err);
      });
  }

  parsePostContent(content) {
    function parseString(c, idx) {
      if (c.includes("http")) {
        // Handle links
        return (
          <span
            style={{ wordBreak: "initial", display: "block", margin: "10px 0" }}
            key={idx}
            className="CuratorFeed--post-link"
          >
            <a href={c} target="_blank" rel="noopener noreferrer">
              {c}
            </a>
            &nbsp;
          </span>
        );
      } else if (c.startsWith("@")) {
        // Handle users
        return (
          <span
            style={{ wordBreak: "initial" }}
            key={idx}
            className="CuratorFeed--user-link"
          >
            <a
              href={`https://twitter.com/${c
                .substring(1, c.length)
                .toLowerCase()}`}
              target="_blank"
              rel="noopener noreferrer"
            >
              {c}
            </a>
            &nbsp;
          </span>
        );
      } else if (c.startsWith("#")) {
        // Handle hashtags
        return (
          <span
            style={{ wordBreak: "initial" }}
            key={idx}
            className="CuratorFeed--hashtag-link"
          >
            <a
              href={`https://twitter.com/search?q=%23${c
                .substring(1, c.length)
                .toLowerCase()}`}
              target="_blank"
              rel="noopener noreferrer"
            >
              {c}
            </a>
            &nbsp;
          </span>
        );
      } else {
        // Handle plain text
        return <Fragment key={idx}>{c} </Fragment>;
      }
    }

    return (
      content &&
      content.length &&
      content.split(" ").map((c, idx) => {
        if (/\r|\n/.exec(c)) {
          return null;
        } else {
          return parseString(c, idx);
        }
      })
    );
  }

  render() {
    const { posts } = this.state;

    return (
      <ul className="CuratorFeed">
        {!posts.length && <p>Loading...</p>}
        {posts &&
          posts.map(p => {
            return (
              <li key={p.id} style={{ listStyleType: "none" }}>
                <div>
                  <div
                    className="CuratorFeed__meta"
                    style={{ display: "flex", alignItems: "center" }}
                  >
                    <img
                      alt={p.user_full_name}
                      src={p.user_image}
                      style={{ marginRight: 10 }}
                    />
                    <div
                      className="CuratorFeed__meta-user"
                      style={{ display: "flex", flexDirection: "column" }}
                    >
                      <span className="CuratorFeed__post-date">
                        {format(new Date(p.source_created_at), "MMM D")}
                      </span>
                      <a
                        rel="noopener noreferrer"
                        style={{ textDecoration: "none" }}
                        href={p.user_url}
                        target="_blank"
                      >
                        @{p.user_full_name}
                      </a>
                    </div>
                  </div>
                  <div className="CuratorFeed__post">
                    <p>{this.parsePostContent(p.text)}</p>
                  </div>
                </div>
              </li>
            );
          })}
      </ul>
    );
  }
}

export default App;
