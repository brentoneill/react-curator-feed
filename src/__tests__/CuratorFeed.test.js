import React from 'react';
import { shallow } from 'enzyme';
import CuratorFeed from '../CuratorFeed';

const config = {
  apiUrl: '',
  apiKey: 'ca1e4636-ca83-4c0a-9877-96279db7e523',
  postsToShow: 6
};

describe('CuratorFeed', () => {
  it('should render correctly in "debug" mode', () => {
    const component = shallow(<CuratorFeed debug config={config}/>);

    expect(component).toMatchSnapshot();
  });
});
