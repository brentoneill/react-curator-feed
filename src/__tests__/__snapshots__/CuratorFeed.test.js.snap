// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CuratorFeed should render correctly in "debug" mode 1`] = `
<ul
  className="CuratorFeed"
>
  <p>
    Loading...
  </p>
</ul>
`;
